packages/playwright-core/src/server/screenshotter.ts

```diff
@@ -147,30 +147,38 @@ export class Screenshotter {
       progress.log('  disabled all CSS animations');
     await Promise.all(this._page.frames().map(async frame => {
       await frame.nonStallingEvaluateInExistingContext('(' + (async function(hideCaret: boolean, disableAnimations: boolean) {
-        const styleTag = document.createElement('style');
+        const collectRoots = (root: Document | ShadowRoot, roots: (Document|ShadowRoot)[] = []): (Document|ShadowRoot)[] => {
+          roots.push(root);
+          const walker = document.createTreeWalker(root, NodeFilter.SHOW_ELEMENT);
+          do {
+            const node = walker.currentNode;
+            const shadowRoot = node instanceof Element ? node.shadowRoot : null;
+            if (shadowRoot)
+              collectRoots(shadowRoot, roots);
+          } while (walker.nextNode());
+          return roots;
+        };
+
+        const styleTags: Element[] = [];
         if (hideCaret) {
-          styleTag.textContent = `
-            *:not(#playwright-aaaaaaaaaa.playwright-bbbbbbbbbbb.playwright-cccccccccc.playwright-dddddddddd.playwright-eeeeeeeee) {
-              caret-color: transparent !important;
-            }
-          `;
-          document.documentElement.append(styleTag);
+          for (const root of collectRoots(document)) {
```

Comment:
Can we call `collectRoots()` just once instead of once for caret and once for animations?